README.rst

```diff
@@ -0,0 +1,806 @@
+
+- Grafeas uses a scheme, namespace, name and version in a URL-like string.
+  https://github.com/Grafeas/Grafeas
+
+- Here.com OSRK uses a package manager, name and version field and a colon-
+  separated URL-like string
+  https://github.com/heremaps/oss-review-toolkit
+
+- JFrog XRay uses a scheme, namespace, name and version in a URL-like string
+  https://www.jfrog.com/confluence/display/XRAY/Xray+REST+API#XrayRESTAPI-ComponentIdentifiers
+
+- Libraries.io uses a platform, name and version
+  https://libraries.io/
+
+- OpenShift fabric8 analytics uses ecosystem, name and version
+  https://github.com/fabric8-analytics/
+
+- ScanCode and AboutCode.org use a type, name and version
+  https://github.com/nexB/scancode-toolkit
+
+- SPDX has an appendix for external repository references and uses a type and a
+  locator with a type-specific syntax for component separators in a URL-like
+  string
+  https://github.com/spdx/spdx-spec/blob/master/chapters/3-package-information.md
+
+- versioneye uses a type, name and version
+  https://github.com/versioneye/
+
+Solution
+========
+
+A `purl` or package URL is an attempt to standardize existing approaches to
+reliably identify and locate software packages.
+
+A `purl` is a URL string used to identify and locate a software package in a
+mostly universal and uniform way across programing languages, package managers,
+packaging conventions, tools, APIs and databases.
+
+Such a package URL is useful to reliably reference the same software package
+using a simple and expressive syntax and conventions based on familiar URLs.
+
+
+purl
+~~~~~
+
+`purl` stands for **package URL**.
+
+A `purl` is a URL composed of six components::
+
+    type:namespace/name@version?qualifiers#subpath
+
+Components are separated by a specific character for unambiguous parsing.
+
+The defintion for each components is:
+
+- **type**: the package "type" or package "protocol" such as maven, npm, nuget,
+  gem, pypi, etc. Required.
+- **namespace**: some name prefix such as a Maven groupid, a Docker image owner,
+  a GitHub user or organization. Optional and type-specific.
+- **name**: the name of the package. Required.
+- **version**: the version of the package. Optional.
+- **qualifiers**: extra qualifying data for a package such as an OS,
+  architecture, a distro, etc. Optional and type-specific.
+- **subpath**: extra subpath within a package, relative to the package root.
+  Optional.
+
+
+Components are designed such that they for a hierarchy from the most significant
+on the left left to the least significant components to the right.
+
+
+A `purl` must NOT contain a URL Authority i.e. there is no support for
+`username`, `password`, `host` and `port` components. A `namespace` segment may
+sometimes look like a `host` but its interpretation is specific to a `type`.
+
+
+Some `purl` examples
+~~~~~~~~~~~~~~~~~~~~
+
+::
+
+    bitbucket:birkenfeld/pygments-main@244fd47e07d1014f0aed9c
```

Comment:
`bitbucket` isn't really a consumer though, is it? Like, what kind of package does that specifier refer to?

This is sticky because npm currently consumes specifiers that look almost like this. Specifically, for npm that would be: `bitbucket:birkenfeld/pygments-main#244fd47e07d1014f0aed9c`. 

The full npm specifier, assuming it was an npm package, would be: `pygments-main@bitbucket:birkenfeld/pygments-main#244fd47e07d1014f0aed9c`

I'm wondering how that would encode as a purl, particularly seeing as `#` is already used for a subpath.

Reply:
@iarna bitbucket, github, gitlab are not strictly-speaking providing packages but are de-factor large reservoirs of packagish-things. They provide more than just a a VCS repo and have tickets, release, some API-fetchable metadata, etc which makes these "packagish" enough to join the fray IMHO.
In the case you mentioned you are encoding the whole VCS address as an NPM version and this works beautifully IMHO:
The purl would be:
`npm:pygments-main@bitbucket:birkenfeld/pygments-main%23244fd47e07d1014f0aed9c`

Of note is the percent encoding of the version `#` to avoid parsing an incorrect `subpath` that would otherwise come out ass this ugly mess:
type='npm', name='pygments-main', version='bitbucket:birkenfeld/pygments-main',  subpath='244fd47e07d1014f0aed9c'